From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Noah van der Aa <ndvdaa@gmail.com>
Date: Sun, 8 Aug 2021 19:56:02 +0200
Subject: [PATCH] Add CompostItemEvent and EntityCompostItemEvent


diff --git a/src/main/java/net/minecraft/world/level/block/ComposterBlock.java b/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
index 528e3c1dc9536408ad080d7f7097ea980bce3e21..2737a06e68b59764b241db6e5b063a38800f7336 100644
--- a/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
@@ -219,7 +219,12 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
 
         if (i < 8 && ComposterBlock.COMPOSTABLES.containsKey(itemstack.getItem())) {
             if (i < 7 && !world.isClientSide) {
-                BlockState iblockdata1 = ComposterBlock.addItem(state, (LevelAccessor) world, pos, itemstack);
+                // Paper start
+                BlockState iblockdata1 = ComposterBlock.addItem(state, (LevelAccessor) world, pos, itemstack, player);
+                if (iblockdata1 == null) {
+                    return InteractionResult.PASS;
+                }
+                // Paper start
 
                 world.levelEvent(1500, pos, state != iblockdata1 ? 1 : 0);
                 player.awardStat(Stats.ITEM_USED.get(itemstack.getItem()));
@@ -243,11 +248,16 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
         if (i < 7 && ComposterBlock.COMPOSTABLES.containsKey(itemstack.getItem())) {
             // CraftBukkit start
             double rand = worldserver.getRandom().nextDouble();
-            BlockState iblockdata1 = ComposterBlock.addItem(iblockdata, DummyGeneratorAccess.INSTANCE, blockposition, itemstack, rand);
+            BlockState iblockdata1 = ComposterBlock.addItem(iblockdata, DummyGeneratorAccess.INSTANCE, blockposition, itemstack, rand, entity); // Paper
             if (iblockdata == iblockdata1 || org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, blockposition, iblockdata1).isCancelled()) {
                 return iblockdata;
             }
-            iblockdata1 = ComposterBlock.addItem(iblockdata, (LevelAccessor) worldserver, blockposition, itemstack, rand);
+            // Paper start
+            iblockdata1 = ComposterBlock.addItem(iblockdata, (LevelAccessor) worldserver, blockposition, itemstack, rand, entity);
+            if (iblockdata1 == null) {
+                return iblockdata;
+            }
+            // Paper end
             // CraftBukkit end
 
             itemstack.shrink(1);
@@ -291,16 +301,42 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
     }
 
     static BlockState addItem(BlockState state, LevelAccessor world, BlockPos pos, ItemStack item) {
-        // CraftBukkit start
-        return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble());
+        return ComposterBlock.addItem(state, world, pos, item, null);
+    }
+    // Paper start
+    static BlockState addItem(BlockState state, LevelAccessor world, BlockPos pos, ItemStack item, @Nullable Entity entity) {
+        return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble(), entity);
     }
+    // Paper end
 
     static BlockState addItem(BlockState iblockdata, LevelAccessor generatoraccess, BlockPos blockposition, ItemStack itemstack, double rand) {
+        // Paper start
+        return ComposterBlock.addItem(iblockdata, generatoraccess, blockposition, itemstack, rand, null);
+    }
+
+    static BlockState addItem(BlockState iblockdata, LevelAccessor generatoraccess, BlockPos blockposition, ItemStack itemstack, double rand, @Nullable Entity entity) {
+        // Paper end
         // CraftBukkit end
         int i = (Integer) iblockdata.getValue(ComposterBlock.LEVEL);
         float f = ComposterBlock.COMPOSTABLES.getFloat(itemstack.getItem());
 
-        if ((i != 0 || f <= 0.0F) && rand >= (double) f) {
+        // Paper start
+        com.destroystokyo.paper.event.block.CompostItemEvent event = null;
+        if (generatoraccess != org.bukkit.craftbukkit.util.DummyGeneratorAccess.INSTANCE) {
+            boolean willRaiseLevel = !((i != 0 || f <= 0.0F) && rand >= (double) f); // taken from if conditional below
+            if (entity == null) {
+                event = new com.destroystokyo.paper.event.block.CompostItemEvent(org.bukkit.craftbukkit.block.CraftBlock.at(generatoraccess, blockposition), itemstack.getBukkitStack(), willRaiseLevel);
+            } else {
+                event = new com.destroystokyo.paper.event.entity.EntityCompostItemEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(generatoraccess, blockposition), itemstack.getBukkitStack(), willRaiseLevel);
+            }
+        }
+
+        if (event != null && !event.callEvent()) {
+            return null;
+        }
+
+        if (event != null && !event.willRaiseLevel()) {
+            // Paper end
             return iblockdata;
         } else {
             int j = i + 1;
@@ -444,6 +480,11 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
                 this.changed = true;
                 BlockState iblockdata = ComposterBlock.addItem(this.state, this.level, this.pos, itemstack);
 
+                // Paper start
+                if (iblockdata == null) {
+                    return;
+                }
+                // Paper end
                 this.level.levelEvent(1500, this.pos, iblockdata != this.state ? 1 : 0);
                 this.removeItemNoUpdate(0);
             }
